### PR TITLE
refactor(ASR): rename `Character` ASR node to `String`

### PR DESCRIPTION
**Pull request by kmr-srbh**

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3344,7 +3344,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         }
 
         // general assignments and checks except when it's a
-        // "Character" declaration
+        // "String" declaration
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:3347]: I would probably not rename this, because the word "Character" here is in reference to how it's used in Fortran (i.e. for AST), and not for ASR.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3353,7 +3353,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 ASR::expr_t* kind_expr = ASRUtils::EXPR(tmp);
                 a_kind = ASRUtils::extract_kind<SemanticError>(kind_expr, sym_type->m_kind->loc);
             }
-            // kind=* only allowed for "Character"
+            // kind=* only allowed for "String"
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:3356]: same here as well I think, it probably is better kept as "Character"

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 3]
```diff
@@ -196,7 +196,7 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
                         break;
                     }
                     default: {
-                        throw SemanticError("Only Character type is supported as of now.", data->type->base.loc);
+                        throw SemanticError("Only String type is supported as of now.", data->type->base.loc);
```
Comment (gxyd) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:199]: this change is good though, as it refers the String in ASR (not AST)

File: tests/reference/asr-array_01_pack-6c829d9.stdout [hunk 4]
```diff
@@ -131,36 +131,8 @@
                             (Allocatable
                                 (Array
                                     (Integer 4)
-                                    [((IntegerConstant 1 (Integer 4) Decimal)
```
Comment (gxyd) [inline tests/reference/asr-array_01_pack-6c829d9.stdout:134]: Sorry, why is this change being created?

File: tests/reference/asr-bits_05-1b98f97.stdout [hunk 5]
```diff
@@ -844,12 +844,12 @@
                                     (Integer 4)
                                     [((IntegerConstant 1 (Integer 4) Decimal)
                                     (IntegerConstant 3 (Integer 4) Decimal))]
-                                    FixedSizeArray
```
Comment (gxyd) [inline tests/reference/asr-bits_05-1b98f97.stdout:847]: I'm not sure why this change is being created, probably when updating the tests. But I don't understand if this is a correct change here.

**Review by gxyd — COMMENTED** [review]
I see that in reference tests (for `*.stdout` files), there are updates other than renaming of `Character` to `String`, I'm not sure if I'll go ahead with those changes unless I'm sure if I understand those.

**Comment by gxyd** [comment]
I believe it's possible that there are incorrect changes being made by *simplifier pass* branch, because of which incorrect reference tests updates might be cause, that's why I'll be a little careful with updating them unless you know what's happening with those.

**Comment by kmr-srbh** [comment]
The updated test references are due to the changes made in the `simplifier` pass. We have quite some of them apart from the ones from `Character` to `String` update. I think we need to update the references for the `Character` to `String` update, but undoing the other changes is difficult give the sheer volume. I have made the best efforts to keep away failing error references. 

I think we can keep the references as they reflect the changes in the `simplifier` pass, which in it's current state is correct for majority of the tests (the CI is green). Thoughts?

**Comment by gxyd** [comment]
> I think we can keep the references as they reflect the changes in the simplifier pass, which in it's current state is correct for majority of the tests (the CI is green). Thoughts?

I appreciate all the work done with this PR and the one against *main* as well. My opinion is that the reference tests update need to be closely looked at (even a small ASR update like `FixedSizeArray` -> `DescriptorArray` may not necessarily be correct, it might reveal an issue with our changes in *simplifier pass*, I'm pretty sure it's possible, and and easy to overlook).

The change `Character` node to `String` node albeit important, is better done later than now, as it'll be a much easy update of references once *simplifier pass* work is merged in with the *main*, I'll try to make sure that's eventually done.

Ofcourse, we'll consult @certik 's opinion on this as well.

**Comment by certik** [comment]
@gxyd if you think it will be too much trouble, then let's do it later.

The priority is to merge the simplifier pass.

**Comment by gxyd** [comment]
> @gxyd if you think it will be too much trouble, then let's do it later.
> 
> The priority is to merge the simplifier pass.

I'll prefer that, it will be much cleaner to look at the changes once the *simplifier pass* is already merged, and I don't see this refactoring necessarily needed right away.

**Comment by certik** [comment]
Ok, let's do it later. @kmr-srbh I apologize for not making this decision sooner. We'll resurrect it once the simplifier branch is merged.
